File: python/aethernet/market/market_service.py
# SPDX-License-Identifier: MIT
"""Offline-capable P2P marketplace (aether-market Phase-2 extension).

Python port of AetherNet.Market.IMarketService / InMemoryMarketService and the listing/escrow models.
Listings are geo-pinned (distributed via aether-space) and may carry a VaultManifest escrow for
document-backed sales; trades run a two-party confirm state machine. Requires aether-space and
aether-vault.
"""
from __future__ import annotations

import uuid
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from enum import IntEnum
from typing import Callable, Dict, List, Optional

from aethernet.market.pov_token import PoVScore
from aethernet.vault.service import VaultManifest


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


class MarketCategory(IntEnum):
    """Category of a :class:`MarketListing`."""

    Goods = 0
    Services = 1
    Labour = 2
    Land = 3
    Documents = 4


class TradeRole(IntEnum):
    """Role of the node confirming a trade step."""

    Buyer = 0
    Seller = 1


class TradeState(IntEnum):
    """State machine for a :class:`TradeEscrow`."""

    Initiated = 0
    BuyerConfirmed = 1
    SellerConfirmed = 2
    Complete = 3
    Disputed = 4


@dataclass
class MarketListing:
    """A geo-pinned market listing dropped by a verified seller. May include a VaultManifest escrow for
    document-backed sales (land deeds, certificates)."""

    listing_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    seller_uhid: str = ""
    seller_pov_score: PoVScore = field(default_factory=PoVScore)
    title: str = ""
    description: str = ""
    price_zar: float = 0.0  # South African Rand
    geohash: str = ""  # 6-char geohash of the listing location
    category: MarketCategory = MarketCategory.Goods
    escrow_manifest: Optional[VaultManifest] = None  # optional Vault escrow
    created_at_utc: datetime = field(default_factory=_now_utc)
    expires_at_utc: datetime = field(default_factory=lambda: _now_utc() + timedelta(days=30))

    @property
    def is_expired(self) -> bool:
        return _now_utc() >= self.expires_at_utc


@dataclass
class TradeEscrow:
    """Tracks the lifecycle of a marketplace trade."""

    escrow_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    listing_id: str = ""
    buyer_uhid: str = ""
    seller_uhid: str = ""
    state: TradeState = TradeState.Initiated
    vault_manifest: Optional[VaultManifest] = None
    created_at_utc: datetime = field(default_factory=_now_utc)


class IMarketService(ABC):
    """The offline-capable P2P marketplace."""

    @abstractmethod
    async def create_listing(self, seller_uhid: str, title: str, description: str, price_zar: float,
                             geohash: str, category: MarketCategory) -> MarketListing: ...

    @abstractmethod
    async def browse_nearby(self, center_geohash: str, radius_cells: int = 2) -> List[MarketListing]: ...

    @abstractmethod
    async def search(self, query: str, category: Optional[MarketCategory] = None) -> List[MarketListing]: ...

    @abstractmethod
    async def initiate_trade(self, listing: MarketListing, buyer_uhid: str) -> TradeEscrow: ...

    @abstractmethod
    async def confirm_trade(self, escrow: TradeEscrow, role: TradeRole) -> TradeEscrow: ...

    @abstractmethod
    async def dispute(self, escrow: TradeEscrow, reason: str) -> None: ...


class InMemoryMarketService(IMarketService):
    """In-memory IMarketService for testing / single-node use; state lost on restart."""

    def __init__(self) -> None:
        self._listings: Dict[str, MarketListing] = {}
        self._escrows: Dict[str, TradeEscrow] = {}
        self.on_listing_received: Optional[Callable[[MarketListing], None]] = None

    async def create_listing(self, seller_uhid: str, title: str, description: str, price_zar: float,
                             geohash: str, category: MarketCategory) -> MarketListing:
        now = _now_utc()
        listing = MarketListing(
            seller_uhid=seller_uhid,
            title=title,
            description=description,
            price_zar=price_zar,
            geohash=geohash,
            category=category,
            created_at_utc=now,
            expires_at_utc=now + timedelta(days=30),
        )
        self._listings[listing.listing_id] = listing
        if self.on_listing_received is not None:
            self.on_listing_received(listing)
        return listing

    async def browse_nearby(self, center_geohash: str, radius_cells: int = 2) -> List[MarketListing]:
        prefix_len = max(1, len(center_geohash) - radius_cells + 1)
        prefix = center_geohash[: min(prefix_len, len(center_geohash))].lower()
        return [
            l for l in self._listings.values()
            if not l.is_expired and l.geohash.lower().startswith(prefix)
        ]

    async def search(self, query: str, category: Optional[MarketCategory] = None) -> List[MarketListing]:
        q = query.lower()
        results: List[MarketListing] = []
        for l in self._listings.values():
            if l.is_expired:
                continue
            if category is not None and l.category != category:
                continue
            if q in l.title.lower() or q in l.description.lower():
                results.append(l)
        return results

    async def initiate_trade(self, listing: MarketListing, buyer_uhid: str) -> TradeEscrow:
        escrow = TradeEscrow(
            listing_id=listing.listing_id,
            buyer_uhid=buyer_uhid,
            seller_uhid=listing.seller_uhid,
            state=TradeState.Initiated,
            vault_manifest=listing.escrow_manifest,
        )
        self._escrows[escrow.escrow_id] = escrow
        return escrow

    async def confirm_trade(self, escrow: TradeEscrow, role: TradeRole) -> TradeEscrow:
        if role == TradeRole.Buyer:
            escrow.state = TradeState.BuyerConfirmed
        else:
            escrow.state = (
                TradeState.Complete if escrow.state == TradeState.BuyerConfirmed
                else TradeState.SellerConfirmed
            )
        self._escrows[escrow.escrow_id] = escrow
        return escrow

    async def dispute(self, escrow: TradeEscrow, reason: str) -> None:
        escrow.state = TradeState.Disputed
        self._escrows[escrow.escrow_id] = escrow

File: kotlin/src/main/kotlin/aethernet/market/MarketService.kt
// SPDX-License-Identifier: MIT
//
// Offline-capable P2P marketplace (aether-market Phase-2 extension). Kotlin port of
// AetherNet.Market.IMarketService / InMemoryMarketService and the listing/escrow models. Listings are
// geo-pinned (distributed via aether-space) and may carry a VaultManifest escrow for document-backed
// sales; trades run a two-party confirm state machine. Requires aether-space and aether-vault.

package aethernet.market

import aethernet.vault.VaultManifest
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/** Category of a [MarketListing]. */
enum class MarketCategory(val value: Byte) {
    Goods(0), Services(1), Labour(2), Land(3), Documents(4);

    companion object {
        fun fromValue(value: Byte): MarketCategory? = entries.find { it.value == value }
    }
}

/** Role of the node confirming a trade step. */
enum class TradeRole(val value: Byte) { Buyer(0), Seller(1) }

/** State machine for a [TradeEscrow]. */
enum class TradeState(val value: Byte) {
    Initiated(0), BuyerConfirmed(1), SellerConfirmed(2), Complete(3), Disputed(4)
}

/**
 * A geo-pinned market listing dropped by a verified seller. May include a [VaultManifest] escrow for
 * document-backed sales (land deeds, certificates).
 */
data class MarketListing(
    val listingId: String,
    val sellerUhid: String,
    val sellerPoVScore: PoVScore? = null,
    val title: String,
    val description: String,
    val priceZAR: Double, // South African Rand
    val geoHash: String,  // 6-char geohash of the listing location
    val category: MarketCategory,
    val escrowManifest: VaultManifest? = null,
    val createdAtUnixMs: Long,
    val expiresAtUnixMs: Long,
) {
    /** Whether the listing has reached its expiry. */
    val isExpired: Boolean get() = System.currentTimeMillis() >= expiresAtUnixMs
}

/** Tracks the lifecycle of a marketplace trade. */
data class TradeEscrow(
    val escrowId: String,
    val listingId: String,
    val buyerUhid: String,
    val sellerUhid: String,
    var state: TradeState,
    val vaultManifest: VaultManifest? = null,
    val createdAtUnixMs: Long,
)

/** The offline-capable P2P marketplace. */
interface MarketService {
    fun createListing(sellerUhid: String, title: String, description: String, priceZAR: Double,
                      geoHash: String, category: MarketCategory): MarketListing
    fun browseNearby(centerGeoHash: String, radiusCells: Int = 2): List<MarketListing>
    fun search(query: String, category: MarketCategory? = null): List<MarketListing>
    fun initiateTrade(listing: MarketListing, buyerUhid: String): TradeEscrow
    fun confirmTrade(escrow: TradeEscrow, role: TradeRole): TradeEscrow
    fun dispute(escrow: TradeEscrow, reason: String)
}

/** In-memory [MarketService] for testing / single-node use; state lost on restart. */
class InMemoryMarketService : MarketService {
    private val listings = ConcurrentHashMap<String, MarketListing>()
    private val escrows = ConcurrentHashMap<String, TradeEscrow>()

    /** Fired when a new listing is received from the mesh or created locally. */
    var onListingReceived: ((MarketListing) -> Unit)? = null

    private companion object {
        const val THIRTY_DAYS_MS = 30L * 24 * 60 * 60 * 1000
    }

    override fun createListing(sellerUhid: String, title: String, description: String, priceZAR: Double,
                              geoHash: String, category: MarketCategory): MarketListing {
        val now = System.currentTimeMillis()
        val listing = MarketListing(
            listingId = UUID.randomUUID().toString(),
            sellerUhid = sellerUhid,
            title = title,
            description = description,
            priceZAR = priceZAR,
            geoHash = geoHash,
            category = category,
            createdAtUnixMs = now,
            expiresAtUnixMs = now + THIRTY_DAYS_MS,
        )
        listings[listing.listingId] = listing
        onListingReceived?.invoke(listing)
        return listing
    }

    override fun browseNearby(centerGeoHash: String, radiusCells: Int): List<MarketListing> {
        val prefixLen = minOf(centerGeoHash.length, maxOf(1, centerGeoHash.length - radiusCells + 1))
        val prefix = centerGeoHash.substring(0, prefixLen).lowercase()
        return listings.values.filter { !it.isExpired && it.geoHash.lowercase().startsWith(prefix) }
    }

    override fun search(query: String, category: MarketCategory?): List<MarketListing> {
        val q = query.lowercase()
        return listings.values.filter {
            !it.isExpired &&
                (category == null || it.category == category) &&
                (it.title.lowercase().contains(q) || it.description.lowercase().contains(q))
        }
    }

    override fun initiateTrade(listing: MarketListing, buyerUhid: String): TradeEscrow {
        val escrow = TradeEscrow(
            escrowId = UUID.randomUUID().toString(),
            listingId = listing.listingId,
            buyerUhid = buyerUhid,
            sellerUhid = listing.sellerUhid,
            state = TradeState.Initiated,
            vaultManifest = listing.escrowManifest,
            createdAtUnixMs = System.currentTimeMillis(),
        )
        escrows[escrow.escrowId] = escrow
        return escrow
    }

    override fun confirmTrade(escrow: TradeEscrow, role: TradeRole): TradeEscrow {
        escrow.state = if (role == TradeRole.Buyer) {
            TradeState.BuyerConfirmed
        } else if (escrow.state == TradeState.BuyerConfirmed) {
            TradeState.Complete
        } else {
            TradeState.SellerConfirmed
        }
        escrows[escrow.escrowId] = escrow
        return escrow
    }

    override fun dispute(escrow: TradeEscrow, reason: String) {
        escrow.state = TradeState.Disputed
        escrows[escrow.escrowId] = escrow
    }
}
